### PR TITLE
ActionSpace

### PR DESCRIPTION
**Pull request by Leuxll**
Essentially the same changes from the deleted ActionSpace branch, but due to divergence for around 25 commits, the best way to resolve this was to save files and delete the branch and remake a new one.

### CODE REVIEW

**Review by derrcodes — APPROVED** [review]
Great. Seems reasonable to make a new branch due to the divergence issue. Everyone should try to keep in mind not to dump a lot of changes (commits) at the same time to prevent this issue.

File: src/main/java/Entities/ActionSpace.java [hunk 1]
```diff
@@ -1,21 +1,127 @@
...
+    private final HashMap<String, ArrayList<Object>> comChestCards;
+    private final HashMap<String, ArrayList<Object>> chanceCards;
+    private final String type;
 
-    public ActionSpace() {
+    public ActionSpace(String type) throws IOException {
+//       /* Assigning type os that when we land on these corresponding cells we can just get the type and draw from the
+//       corresponding cell*/
+        this.type = type;
+//       /* Initialising the different Card HashMaps as we want to set these to the private instance attributes later*/
```
Comment (benxu3) [inline src/main/java/Entities/ActionSpace.java:18]: i think it would be better design if you made this code in to a helper method

**Review by benxu3 — CHANGES_REQUESTED** [review]

**Review by benxu3 — APPROVED** [review]

**Review by kRichard32 — APPROVED** [review]
In the future, we should move the load function outside of this entity and into the use case interactor in order to oblige to clean architecture. This change is not need right now; we can move it later once the pull request for the game logic interactor and the game creation interactor has completed. Other than that, I was wondering why we need to alter the performAction method inside the cell class.

**Review by kRichard32 — CHANGES_REQUESTED** [review]
We might want to store the type of actionSpace inside the class itself instead of having it be passed through the performAction parameter

**Comment by Leuxll** [comment]
> In the future, we should move the load function outside of this entity and into the use case interactor in order to oblige to clean architecture. This change is not need right now; we can move it later once the pull request for the game logic interactor and the game creation interactor has completed. Other than that, I was wondering why we need to alter the performAction method inside the cell class. I was thinking we could store the type of ActionSpace object inside the ActionSpace itself, rather than have the type be passed by parameter. Other than these two concerns the code looks logically done and neat.

I think the point was that cell could either be a property or an action space, so we made it into an abstract class that would be used when the board is being built, but I was unclear on that as well. As for storing the type within ActionSpace, it is already a private instance attribute so I am not quite sure what you are referring to.

**Comment by kRichard32** [comment]
Yes, using a getter method for the type rather than having the performAction method have to pass a parameter indicating the type would be better.

**Comment by Leuxll** [comment]
@kRichard32 In terms of code smell, I think the methods are a little bit bloated so I might just refactor some of the code into helper functions as well.

**Review by Leuxll — COMMENTED** [review]
@kRichard32 Removed the passing of actionType within perform action method and also updated the corresponding methods in cell and also property.

**Review by derrcodes — APPROVED** [review]
Just looked at the changes. I agree with the changes cause it makes everything less cluttered. Making changes on each of those may also be easier later on in case.

**Review by kRichard32 — APPROVED** [review]
With those changes, it looks good to me. Good work!

File: src/main/java/Entities/Cell.java [hunk 2]
```diff
@@ -1,7 +1,7 @@
 package Entities;
 
 abstract class Cell {
-    public String performAction(){
-        return "";
-    }
+
+    public abstract String performAction(Player player, Board board);
```
Comment (benxu3) [inline src/main/java/Entities/Cell.java:5]: I called the player in my Property PR to currentPlayer and merged my PR to main already. Could you change this to parameter to currentPlayer? This way we will not have any merge conflicts.

**Review by benxu3 — COMMENTED** [review]
Take a look at my comments, otherwise I think it is good to merge. Thanks for writing JavaDoc and Unit tests.

**Pull request merged**